File: a2ui/catalog.go
package a2ui

// Catalog describes the set of components and functions a surface may use. The
// envelope schema is catalog-agnostic; a Catalog supplies the information needed
// to validate component types, function names and parent/child structure.
type Catalog struct {
	// ID is the unique catalog identifier (matches CreateSurface.CatalogID).
	ID string
	// Components maps component type names to their definitions.
	Components map[string]ComponentDef
	// Functions is the set of registered function names.
	Functions map[string]struct{}
	// ThemeProperties is the set of recognized theme property names.
	ThemeProperties map[string]struct{}
}

// ComponentDef describes a single component type within a catalog, including
// which of its properties hold structural references to other components.
type ComponentDef struct {
	// Name is the component type name.
	Name string
	// ChildFields lists properties holding a single child component ID.
	ChildFields []string
	// ChildListFields lists properties holding a ChildList (array of IDs or a
	// template).
	ChildListFields []string
	// NestedChildFields maps property names (containing an array of objects) to
	// the key within each object that holds a child component ID. This handles
	// components like Tabs where children are nested inside a structured array.
	NestedChildFields map[string]string
	// Input reports whether the component is an interactive input (establishing
	// two-way data binding).
	Input bool
}

// HasComponent reports whether the catalog defines the given component type.
func (c *Catalog) HasComponent(typ string) bool {
	_, ok := c.Components[typ]
	return ok
}

// HasFunction reports whether the catalog registers the given function name.
func (c *Catalog) HasFunction(name string) bool {
	_, ok := c.Functions[name]
	return ok
}

// BasicCatalogID is the identifier of the A2UI v0.9.1 basic component catalog.
const BasicCatalogID = "https://a2ui.org/specification/v0_9_1/catalogs/basic/catalog.json"

// BasicCatalog returns the standard A2UI v0.9.1 basic catalog: the baseline set
// of components, functions and theme properties.
func BasicCatalog() *Catalog {
	funcs := []string{
		"required", "regex", "length", "numeric", "email",
		"formatString", "formatNumber", "formatCurrency", "formatDate",
		"pluralize", "openUrl", "and", "or", "not",
	}
	fnSet := make(map[string]struct{}, len(funcs))
	for _, f := range funcs {
		fnSet[f] = struct{}{}
	}

	themeProps := []string{"primaryColor", "iconUrl", "agentDisplayName"}
	themeSet := make(map[string]struct{}, len(themeProps))
	for _, p := range themeProps {
		themeSet[p] = struct{}{}
	}

	defs := []ComponentDef{
		{Name: "Text"},
		{Name: "Image"},
		{Name: "Icon"},
		{Name: "Video"},
		{Name: "AudioPlayer"},
		{Name: "Row", ChildListFields: []string{"children"}},
		{Name: "Column", ChildListFields: []string{"children"}},
		{Name: "List", ChildListFields: []string{"children"}},
		{Name: "Card", ChildFields: []string{"child"}},
		{Name: "Tabs", NestedChildFields: map[string]string{"tabs": "child"}},
		{Name: "Divider"},
		{Name: "Modal", ChildFields: []string{"child", "entryPointChild"}},
		{Name: "Button", ChildFields: []string{"child"}},
		{Name: "CheckBox", Input: true},
		{Name: "TextField", Input: true},
		{Name: "DateTimeInput", Input: true},
		{Name: "ChoicePicker", Input: true},
		{Name: "Slider", Input: true},
	}
	comps := make(map[string]ComponentDef, len(defs))
	for _, d := range defs {
		comps[d.Name] = d
	}

	return &Catalog{
		ID:              BasicCatalogID,
		Components:      comps,
		Functions:       fnSet,
		ThemeProperties: themeSet,
	}
}

// CatalogRegistry maps catalog IDs to catalogs, enabling validation against the
// catalog declared by each surface.
type CatalogRegistry struct {
	catalogs map[string]*Catalog
}

// NewCatalogRegistry returns a registry pre-populated with the basic catalog.
func NewCatalogRegistry() *CatalogRegistry {
	r := &CatalogRegistry{catalogs: map[string]*Catalog{}}
	r.Register(BasicCatalog())
	return r
}

// Register adds (or replaces) a catalog in the registry.
func (r *CatalogRegistry) Register(c *Catalog) {
	if c == nil {
		return
	}
	r.catalogs[c.ID] = c
}

// Lookup returns the catalog for the given ID, if registered.
func (r *CatalogRegistry) Lookup(id string) (*Catalog, bool) {
	c, ok := r.catalogs[id]
	return c, ok
}

File: desktop/frontend/src/a2ui-renderer/catalog.ts
/**
 * A2UI CatalogRegistry + BasicCatalog（对齐 Go a2ui/catalog.go）。
 *
 * 定义组件类型、函数名、主题属性的注册表。
 * BasicCatalog 覆盖 A2UI v0.9.1 的 18 个标准组件和 15 个内置函数。
 */

// ── 类型 ───────────────────────────────────────────

/** 单个组件类型的定义。 */
export interface ComponentDef {
  /** 组件类型名（如 "Text", "Button", "Column"）。 */
  name: string
  /** 持有单个子组件 ID 的属性名。 */
  childFields: string[]
  /** 持有子组件 ID 列表的属性名。 */
  childListFields: string[]
  /**
   * 嵌套子组件映射：属性名 → 子组件 ID 所在的键名。
   * 用于 Tabs 等组件，其 children 嵌套在结构化数组中。
   */
  nestedChildFields: Record<string, string>
  /** 是否为交互式输入组件（建立双向数据绑定）。 */
  input: boolean
}

/** A2UI Catalog。 */
export interface Catalog {
  /** 唯一 catalog ID（匹配 CreateSurface.CatalogID）。 */
  id: string
  /** 组件类型 → 组件定义。 */
  components: Record<string, ComponentDef>
  /** 已注册的函数名集合。 */
  functions: Set<string>
  /** 已识别的主题属性名集合。 */
  themeProperties: Set<string>
}

// ── 辅助函数 ───────────────────────────────────────

function _setOf(items: string[]): Set<string> {
  return new Set(items)
}

// ── BasicCatalog ───────────────────────────────────

export const BasicCatalogID =
  'https://a2ui.org/specification/v0_9_1/catalogs/basic/catalog.json'

/** 返回 A2UI v0.9.1 BasicCatalog。 */
export function basicCatalog(): Catalog {
  const funcs = [
    'required', 'regex', 'length', 'numeric', 'email',
    'formatString', 'formatNumber', 'formatCurrency', 'formatDate',
    'pluralize', 'openUrl', 'and', 'or', 'not',
  ]

  const themeProps = ['primaryColor', 'iconUrl', 'agentDisplayName']

  const defs: ComponentDef[] = [
    { name: 'Text', childFields: [], childListFields: [], nestedChildFields: {}, input: false },
    { name: 'Image', childFields: [], childListFields: [], nestedChildFields: {}, input: false },
    { name: 'Icon', childFields: [], childListFields: [], nestedChildFields: {}, input: false },
    { name: 'Video', childFields: [], childListFields: [], nestedChildFields: {}, input: false },
    { name: 'AudioPlayer', childFields: [], childListFields: [], nestedChildFields: {}, input: false },
    { name: 'Row', childFields: [], childListFields: ['children'], nestedChildFields: {}, input: false },
    { name: 'Column', childFields: [], childListFields: ['children'], nestedChildFields: {}, input: false },
    { name: 'List', childFields: [], childListFields: ['children'], nestedChildFields: {}, input: false },
    { name: 'Card', childFields: ['child'], childListFields: [], nestedChildFields: {}, input: false },
    { name: 'Tabs', childFields: [], childListFields: [], nestedChildFields: { tabs: 'child' }, input: false },
    { name: 'Divider', childFields: [], childListFields: [], nestedChildFields: {}, input: false },
    { name: 'Modal', childFields: ['child', 'entryPointChild'], childListFields: [], nestedChildFields: {}, input: false },
    { name: 'Button', childFields: ['child'], childListFields: [], nestedChildFields: {}, input: false },
    { name: 'CheckBox', childFields: [], childListFields: [], nestedChildFields: {}, input: true },
    { name: 'TextField', childFields: [], childListFields: [], nestedChildFields: {}, input: true },
    { name: 'DateTimeInput', childFields: [], childListFields: [], nestedChildFields: {}, input: true },
    { name: 'ChoicePicker', childFields: [], childListFields: [], nestedChildFields: {}, input: true },
    { name: 'Slider', childFields: [], childListFields: [], nestedChildFields: {}, input: true },
  ]

  const comps: Record<string, ComponentDef> = {}
  for (const d of defs) {
    comps[d.name] = d
  }

  return {
    id: BasicCatalogID,
    components: comps,
    functions: _setOf(funcs),
    themeProperties: _setOf(themeProps),
  }
}

// ── CatalogRegistry ────────────────────────────────

/** Catalog 注册表，支持按 ID 查找。 */
export class CatalogRegistry {
  private _catalogs = new Map<string, Catalog>()

  constructor() {
    this.register(basicCatalog())
  }

  /** 注册（或替换）一个 Catalog。 */
  register(cat: Catalog): void {
    this._catalogs.set(cat.id, cat)
  }

  /** 按 ID 查找 Catalog。 */
  lookup(id: string): Catalog | undefined {
    return this._catalogs.get(id)
  }
}
